src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProvider.Ide.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.CodingConventions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Options
+{
+    internal sealed partial class EditorConfigDocumentOptionsProvider
+    {
+        private static readonly ConditionalWeakTable<CodingConventionsChangedEventArgs, HashSet<ProjectId>> s_projectNotifications =
+            new ConditionalWeakTable<CodingConventionsChangedEventArgs, HashSet<ProjectId>>();
+
+        partial void OnCodingConventionContextCreated(DocumentId documentId, ICodingConventionContext context)
+        {
+            context.CodingConventionsChangedAsync += (sender, e) => HandleCodingConventionsChangedAsync(documentId, sender, e);
+        }
+
+        private Task HandleCodingConventionsChangedAsync(DocumentId documentId, object sender, CodingConventionsChangedEventArgs e)
+        {
+            var projectId = documentId.ProjectId;
+            var projectsAlreadyNotified = s_projectNotifications.GetOrCreateValue(e);
+            lock (projectsAlreadyNotified)
+            {
+                if (!projectsAlreadyNotified.Add(projectId))
+                {
+                    return Task.CompletedTask;
+                }
+            }
+
+            var foregroundNotificationService = ((IMefHostExportProvider)_workspace.Services.HostServices).GetExports<IForegroundNotificationService>().Single().Value;
+            foregroundNotificationService.RegisterNotification(UpdateProject, _listener.BeginAsyncOperation(nameof(HandleCodingConventionsChangedAsync)), CancellationToken.None);
+
+            return Task.CompletedTask;
+
+            void UpdateProject()
+            {
+                var compilationOptions = _workspace.CurrentSolution.GetProject(projectId)?.CompilationOptions;
+                if (compilationOptions == null)
+                {
+                    return;
+                }
+
+                var parseOptions = _workspace.CurrentSolution.GetProject(projectId).ParseOptions;
```

Comment:
Once my free-threaded workspace PR merges, this won't be safe: the project could have been removed between the fetch of the CompilationOptions and ParseOptions.

Reply:
❓ Is the free-threaded workspace targeting 15.8?

❓ What is the new atomic workspace update API in the free-threaded implementation?

Reply:
We're at least going to be testing internally in 15.8.

The only API that's available for a _client_ of the workspace to update it is TryApplyChanges. Random other clients changing stuff like this is intentionally not supported.

Reply:
Basically: the workspace is designed to reflect the reality known by the project system: "I reject your reality and substitute my own" isn't in the design.